src/Tracer.php

```diff
@@ -0,0 +1,203 @@
+    {
+        /** @var TestMethod */
+        $testMethod = $event->test();
+
+        $this->result['tests'][] = new Result(
+            $testMethod->testDox()->prettifiedMethodName(),
+            'pass',
+            $this->methodCode($testMethod),
+            $this->taskId($testMethod),
+        );
+    }
+
+    private function addTestFailed(Failed $event): void
+    {
+        /** @var TestMethod */
+        $testMethod = $event->test();
+
+        $phpUnitMessage = \trim($event->throwable()->asString());
+        $phpUnitMessage = \str_replace(
+            $this->exerciseDir . '/',
+            '',
+            $phpUnitMessage
+        );
+        $phpUnitMessage = $testMethod->nameWithClass() . "\n" . $phpUnitMessage;
+
+        $this->result['tests'][] = new Result(
+            $testMethod->testDox()->prettifiedMethodName(),
+            'fail',
+            $this->methodCode($testMethod),
+            $this->taskId($testMethod),
+            '',
+            $phpUnitMessage,
+        );
+    }
+
+    private function addTestErrored(Errored $event): void
+    {
+        /** @var TestMethod */
+        $testMethod = $event->test();
+
+        $phpUnitMessage = \trim($event->throwable()->asString());
+        $phpUnitMessage = \str_replace(
+            $this->exerciseDir . '/',
+            '',
+            $phpUnitMessage
+        );
+        $phpUnitMessage = $testMethod->nameWithClass() . "\n" . $phpUnitMessage;
+
+        $this->result['tests'][] = new Result(
+            $testMethod->testDox()->prettifiedMethodName(),
+            'error',
+            $this->methodCode($testMethod),
+            $this->taskId($testMethod),
+            '',
+            $phpUnitMessage,
+        );
+    }
+
+    private function addBeforeFirstTestMethodErrored(BeforeFirstTestMethodErrored $event): void
+    {
+        $phpUnitMessage = \trim($event->throwable()->asString());
+        $phpUnitMessage = \str_replace(
+            $this->exerciseDir . '/',
+            '',
+            $phpUnitMessage
+        );
+
+        $this->result['status'] = 'error';
+        $this->result['message'] = $phpUnitMessage;
+    }
+
+    private function addTestOutput(PrintedUnexpectedOutput $event): void
+    {
+        // This must rely on the sequence of events!
+
+        /** @var Result */
+        $lastTest = $this->result['tests'][\array_key_last($this->result['tests'])];
+        $lastTest->setUserOutput($event->output());
+    }
+
+    private function saveResults(): void
+    {
+        foreach ($this->result['tests'] as $result) {
+            if ($result->isFailed() || $result->isErrored()) {
+                $this->result['status'] = 'fail';
+            }
+        }
+
+        \file_put_contents(
+            $this->outFileName,
+            \json_encode($this->result) . "\n",
+            // \json_encode($this->result, JSON_PRETTY_PRINT) . "\n",
+        );
+    }
+
+    private function methodCode(TestMethod $testMethod): string
```

Comment:
This feel a bit hacky to me

Reply:
It is :smiley: But there is no "nice" way without using a big dependency. PHPUnit does not provide test method code in the events.